vision.md

```diff
@@ -206,26 +206,188 @@ The framework includes modules for cloud connectivity, data mapping, device
 management, intra-edge communication, and certificate management, all aspects
 and challenges our target personas need to address.
 
+We therefore define the following requirements.
+
+### Non-functional requirements
+
+<!--
+The following list of non-functional requirements should be considered
+non-final. It might be extended in the future.
+-->
+
+1. Multi-paradigm edge framework
+    - Several kinds of users are be able to consume thin-edge.io at different
+      stages (e.g. using the Rust code as a framework to develop their own
+      solutions, using (networking)interfaces like for example MQTT to attach
+      their own infrastructure to thin-edge.io, or using the final
+      deployment-ready binary to solve their problems).
+1. Introspective functionality to enable quality assurance as well as support of
+   deployments
+    - Knowing what happens inside a system is crucial to be able to quickly
+      identify misbehaving components. Users are be able to access an overview
+      of the communication of the different components
+1. Ability to attach data sources and data sinks to thin-edge.io via different
+   mechanisms, either by implementing connectors directly in the framework using
+   rust code, attaching them via some networking mechanism (e.g. MQTT) or other
+   mechanisms the framework provides
+    1. The framework ships with connectivity functionality for the most common
+       connectivity requirements
+    1. Configuration of connectivity functionality is exposed to the user
+    1. Connectivity over the same mechanism but with different configuration
+       settings is easily doable for the user (e.g. using MQTT to connect to
+       different brokers with completely different settings)
+1. Easy deployment and little friction when using thin-edge.io as a framework to
+   develop their own solutions based on it
+1. Trust in the user to do the right thing with the tools provided, respect
+   their decisions and guide them towards a maintainable solution by design
+    - Users are assumed to be experts in their domain and should be given the
+      tools they need to solve the problems they have. Unnecessarily burdening
+      them with abstractions only causes frustration
```

Comment:
> Unnecessarily burdening them with abstractions only causes frustration

Can you give an abstraction that would introduce such a burden?

Reply:
For example hiding away the HTTP/MQTT component to some inner framework abstraction that cannot easily be interacted with. Imagine they would have to set a proxy or so. If we are not careful with what kind of abstraction we are running at, the user might not be able to insert their configuration here and thus be in a powerless situation.